src/style-spec/function/expressions.js

```diff
@@ -0,0 +1,354 @@
+'use strict';
+
+// @flow
+
+const assert = require('assert');
+
+const {
+    NumberType,
+    StringType,
+    BooleanType,
+    ColorType,
+    ObjectType,
+    ValueType,
+    InterpolationType,
+    typename,
+    variant,
+    vector,
+    array,
+    anyArray,
+    lambda,
+    nargs
+} = require('./types');
+
+/*::
+ import type { PrimitiveType, TypeName, VariantType, VectorType, ArrayType, AnyArrayType, NArgs, LambdaType, Type } from './types.js';
+
+ import type { TypeError, TypedLambdaExpression, TypedLiteralExpression, TypedExpression } from './type_check.js';
+
+ import type { CompiledExpression } from './compile.js';
+
+ export type ExpressionName = "literal" | "ln2" | "pi" | "e" | "string" | "number" | "boolean" | "json_array" | "object" | "get" | "has" | "at" | "typeof" | "length" | "zoom" | "properties" | "geometry_type" | "id" | "case" | "match" | "is_error" | "==" | "!=" | ">" | ">=" | "<=" | "<" | "&&" | "||" | "!" | "curve" | "step" | "exponential" | "linear" | "cubic-bezier" | "+" | "-" | "*" | "/" | "%" | "^" | "log10" | "ln" | "log2" | "sin" | "cos" | "tan" | "asin" | "acos" | "atan" | "ceil" | "floor" | "round" | "abs" | "min" | "max" | "concat" | "upcase" | "downcase" | "rgb" | "rgba" | "color" | "color_to_array"
+
+ type Definition = {
+     name: ExpressionName,
+     type: Type,
+     compile: (expr: TypedExpression, args: Array<CompiledExpression>) => ({ js?: string, errors?: Array<string>, isFeatureConstant?: boolean, isZoomConstant?: boolean })
+ }
+ */
+
+const expressions: { [string]: Definition } = {
+    'ln2': defineMathConstant('ln2'),
+    'pi': defineMathConstant('pi'),
+    'e': defineMathConstant('e'),
+    'string': {
+        name: 'string',
+        type: lambda(StringType, ValueType),
+        compile: (_, args) => ({ js: `String(${args[0].js})` })
+    },
+    'number': {
+        name: 'string',
+        type: lambda(NumberType, ValueType),
+        compile: (_, args) => ({js: `Number(${args[0].js})`})
+    },
+    'boolean': {
+        name: 'boolean',
+        type: lambda(BooleanType, ValueType),
+        compile: (_, args) => ({js: `Boolean(${args[0].js})`})
+    },
+    'json_array': {
+        name: 'json_array',
+        type: lambda(vector(ValueType), ValueType),
+        compile: fromContext('asArray')
+    },
+    'object': {
+        name: 'object',
+        type: lambda(ObjectType, ValueType),
+        compile: fromContext('asObject')
+    },
```

Comment:
`json_array` and `object` are assertions (yielding an error if the input isn't the right type), whereas `number`, `string`, `boolean` are coercions.  Should we _also_ have assertions for number/string/boolean?